vault-core/src/main/java/org/apache/jackrabbit/vault/util/CachingSupplier.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.jackrabbit.vault.util;
+
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A simple utility to resolve values lazy
+ * @param <T>
+ */
+
+public class CachingSupplier<T> {
+
+    T resolvedValue = null;
+    Supplier<T> resolver;
+    
+    private CachingSupplier (Supplier<T> supplier) {
+        resolver = supplier;
+    }
+    
+    public static<T> CachingSupplier<T> of(@NotNull Supplier<T> supplier) {
+        return new CachingSupplier<T>(supplier);
+    }
+
+    /**
+     * Resolve the value if necessary and return it; if the resolved value is non-null, it
```

Comment:
Hard to understand.
Maybe something like `Uses the cached value from previous get() calls instead of contacting the underlying resolver`.